modules/optflow/src/pcaflow.cpp

```diff
@@ -0,0 +1,508 @@
+  prior->fillConstraints( A1.ptr<float>( features.size(), 0 ), A2.ptr<float>( features.size(), 0 ),
+                          b1.ptr<float>( features.size(), 0 ), b2.ptr<float>( features.size(), 0 ) );
+}
+
+static void applyCLAHE( UMat &img )
+{
+  Ptr<CLAHE> clahe = createCLAHE();
+  clahe->setClipLimit( 14 );
+  clahe->apply( img, img );
+}
+
+static void reduceToFlow( const Mat &w1, const Mat &w2, Mat &flow, const Size &basisSize )
+{
+  const Size size = flow.size();
+  Mat flowX( size, CV_32F, 0.0f );
+  Mat flowY( size, CV_32F, 0.0f );
+
+  const float mult = sqrt( size.area() ) * 0.5;
+
+  for ( int i = 0; i < basisSize.width; ++i )
+    for ( int j = 0; j < basisSize.height; ++j )
+    {
+      flowX.at<float>( j, i ) = w1.at<float>( i * basisSize.height + j ) * mult;
+      flowY.at<float>( j, i ) = w2.at<float>( i * basisSize.height + j ) * mult;
+    }
+  for ( int i = 0; i < basisSize.height; ++i )
+  {
+    flowX.at<float>( i, 0 ) *= M_SQRT2;
+    flowY.at<float>( i, 0 ) *= M_SQRT2;
+  }
+  for ( int i = 0; i < basisSize.width; ++i )
+  {
+    flowX.at<float>( 0, i ) *= M_SQRT2;
+    flowY.at<float>( 0, i ) *= M_SQRT2;
+  }
+
+  dct( flowX, flowX, DCT_INVERSE );
+  dct( flowY, flowY, DCT_INVERSE );
+  for ( int i = 0; i < size.height; ++i )
+    for ( int j = 0; j < size.width; ++j )
+      flow.at<Point2f>( i, j ) = Point2f( flowX.at<float>( i, j ), flowY.at<float>( i, j ) );
+}
+
+void OpticalFlowPCAFlow::calc( InputArray I0, InputArray I1, InputOutputArray flowOut )
+{
+  const Size size = I0.size();
+  CV_Assert( size == I1.size() );
+
+  UMat from, to;
+  if ( I0.channels() == 3 )
+  {
+    cvtColor( I0, from, COLOR_BGR2GRAY );
+    from.convertTo( from, CV_8U );
+  }
+  else
+  {
+    I0.getMat().convertTo( from, CV_8U );
+  }
+  if ( I1.channels() == 3 )
+  {
+    cvtColor( I1, to, COLOR_BGR2GRAY );
+    to.convertTo( to, CV_8U );
+  }
+  else
+  {
+    I1.getMat().convertTo( to, CV_8U );
+  }
+
+  CV_Assert( from.channels() == 1 );
+  CV_Assert( to.channels() == 1 );
+
+  const Mat fromOrig = from.getMat( ACCESS_READ ).clone();
+
+  applyCLAHE( from );
+  applyCLAHE( to );
+
+  std::vector<Point2f> features, predictedFeatures;
+  findSparseFeatures( from, to, features, predictedFeatures );
+  removeOcclusions( from, to, features, predictedFeatures );
+
+  flowOut.create( size, CV_32FC2 );
+  Mat flow = flowOut.getMat();
+
+  Mat w1, w2;
+  if ( prior.get() )
+  {
+    Mat A1, A2, b1, b2;
+    getSystem( A1, A2, b1, b2, features, predictedFeatures, size );
+    solveLSQR( A1, b1, w1, dampingFactor * size.area() );
+    solveLSQR( A2, b2, w2, dampingFactor * size.area() );
+  }
+  else
+  {
+    Mat A, b1, b2;
+    getSystem( A, b1, b2, features, predictedFeatures, size );
+    solveLSQR( A, b1, w1, dampingFactor * size.area() );
+    solveLSQR( A, b2, w2, dampingFactor * size.area() );
+  }
+  Mat flowSmall( ( size / 8 ) * 2, CV_32FC2 );
```

Comment:
What's the rationale behind this small size?  perhaps a comment, and suggest (size/4)